File: Apps/BankBusiness/backend/repository/PaymentRepository.cs
using backend.dto.payment;
using backend.model;
using Microsoft.EntityFrameworkCore;

namespace backend.repository;

public class PaymentRepository(DatabaseContext _context)
{
    public async Task<Payment> Find(Guid id)
    {
        return await _context.Payments.FindAsync(id);
    }

    public async Task<List<PaymentWithAmountDto>> FindAllByPaymentAccount(long paymentAccountId)
    {
        var t = _context.Payments
            .Where(p => p.ProviderPaymentAccount.Id == paymentAccountId)
            .Select(p => new 
            {
                paymentId = p.Id,
                invoiceId = p.Invoice.Id
            }).ToList();
        var t1 = _context.Invoices
            .Where(i => t.Any(tpl => tpl.invoiceId == i.Id))
            .Select(i => new {i.amount, i.Id})
            .ToList();
        return t.Join(
            t1,
            tpl1 => tpl1.invoiceId,
            tpl2 => tpl2.Id,
            (tpl1, tpl2) => new PaymentWithAmountDto(tpl1.paymentId, tpl2.amount)
        ).ToList();
        // return await _context.Payments.FromSqlRaw($"select p.id from payment p where p.payment_account_id = {paymentAccountId}").ToListAsync();
    }

    public async Task<PaymentWithAmountDto?> FindPaymentById(Guid paymentId)
    {
        var t = await _context.Payments
            .Where(p => p.Id == paymentId)
            .Select(p => new
            {
                paymentId = p.Id,
                invoiceId = p.Invoice.Id
            }).FirstOrDefaultAsync();
        var t1 = _context.Invoices
            .Where(i => i.Id == t!.invoiceId)
            .Select(i => new {i.amount, i.Id})
            .FirstOrDefaultAsync();
        return new PaymentWithAmountDto(t.paymentId, t1.Result!.amount);
    }

    public async Task<bool> CheckPayment(Guid invoiceId)
    {
        return (await _context.Payments.FromSql($"select * from payment p where p.invoice_id = {invoiceId}").CountAsync()) == 1;
    }
    
}

File: Apps/BankBusiness/backend/src/main/scala/dao/repository/PaymentRepository.scala
package dao.repository

import dto.payment.PaymentWithAmountDto
import model.{Invoice, Payment}
import zio.{Task, URLayer, ZLayer}

import java.util.UUID
import javax.sql.DataSource

trait PaymentRepository {
  def findDtoById(id: UUID): Task[Option[PaymentWithAmountDto]]
  def findPaymentAccountIdById(id: UUID): Task[Option[Long]]
  def findAllByPaymentAccountId(paymentAccountId: Long): Task[List[PaymentWithAmountDto]]
  def checkPaymentByInvoiceId(invoiceId: UUID): Task[Boolean]
}

case class PaymentRepositoryLive(private val ds: DataSource) extends PaymentRepository {

  import io.getquill.*
  import dao.DatabaseContext._

  private inline def payments = quote(query[Payment])
  private inline def invoices = quote(query[Invoice])

  private val dsLayer = ZLayer.succeed(ds)

  override def findDtoById(paymentId: UUID): Task[Option[PaymentWithAmountDto]] = {
    run(quote{
      payments
        .join(invoices)
        .on({case (p, i) => i.id == p.invoiceId})
        .filter((p, i) => p.id == lift(paymentId))
        .map((p, i) => (p.id, i.amount))
    })
      .mapBoth(e => Throwable(e.getMessage), _.headOption)
      .map(opt => {
        if opt.isDefined then Option(PaymentWithAmountDto(opt.get._1, opt.get._2))
        else Option.empty[PaymentWithAmountDto]
      })
      .provideLayer(dsLayer)
  }

  override def findPaymentAccountIdById(id: UUID): Task[Option[Long]] = {
    run(quote{
      payments
        .filter(_.id == lift(id))
        .map(_.paymentAccountId)
    })
      .mapBoth(e => Throwable(e.getMessage), _.headOption)
      .provideLayer(dsLayer)
  }

  override def findAllByPaymentAccountId(paymentAccountId: Long): Task[List[PaymentWithAmountDto]] = {
    run(quote {
      payments
        .join(invoices)
        .on((p, i) => p.invoiceId == i.id)
        .filter((p, i) => p.paymentAccountId == lift(paymentAccountId))
        .map((p, i) => (p.id, i.amount))
    }).mapBoth(e => Throwable(e.getMessage), c => c.map({case (id, amount) => PaymentWithAmountDto(id, amount)}))
      .provideLayer(dsLayer)
  }

  override def checkPaymentByInvoiceId(invoiceId: UUID): Task[Boolean] = {
    run(quote{
      payments
        .filter(_.invoiceId == lift(invoiceId))
        .size
    })
      .mapBoth(e => Throwable(e.getMessage), _ == 1)
      .provideLayer(dsLayer)
  }
}

object PaymentRepositoryLive {
  val layer: URLayer[DataSource, PaymentRepositoryLive] = ZLayer.fromFunction(PaymentRepositoryLive.apply _)
}